### PR TITLE
Feature/e UI 2831 multi select dynamic phillip whitaker

### PR DESCRIPTION
**Pull request by phillip-whitaker**
Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-2831

### Change description ###

Create Multiselect checkbox input value

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      fixture = TestBed.createComponent(ReadDynamicMultiSelectListFieldComponent);
+      component = fixture.componentInstance;
+
+      component.caseField = CASE_FIELD;
+
+      de = fixture.debugElement;
+      fixture.detectChanges();
+    }));
+
+    it('render values as a table with one row and one cell per value', () => {
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts:80]: Additionally, can we have some tests for the _other_ `value` structure - the one where it's an object that contains an array of `list_items` and an array of `value` objects?

Comment (paul-graham) [reply src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts:80]: This might actually be only for the write fields... not sure.

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.html [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<table *ngIf="caseField.value && caseField.value.length" class="multi-select-list-field-table">
+    <tbody>
+      <tr *ngFor="let value of caseField.value">
+        <td><span class="text-16">{{ value | ccdFixedList:caseField.field_type.fixed_list_items }}</span></td>
+      </tr>
+    </tbody>
+  </table>
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.html:7]: Indenting peculiarity in here...

File: src/shared/components/palette/dynamic-multi-select-list/multi-select-list.scss [hunk 3]
```diff
@@ -0,0 +1,16 @@
+.multi-select-list-field-table {
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/multi-select-list.scss:1]: Should this be called `read-dynamic-multi-select-list-field.component.scss` since that's the only place it's used?

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    const checkboxes = de.queryAll($CHECKBOXES);
+
+    expect(checkboxes.length).toEqual(FIELD_LIST_ITEMS.length);
+
+    FIELD_LIST_ITEMS.forEach(item => {
+      expect(checkboxes.find(checkbox => attr(checkbox, 'value') === item.code)).toBeTruthy();
+    });
+  });
+
+  it('should mark as selected the initially selected checkboxes', () => {
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.spec.ts:94]: Add a test for when the value is null and/or a single item, rather than all items?

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    this.registerControl(this.checkboxes, true);
+  }
+
+  onCheckChange(event) {
+    if (!this.isSelected(event.target.value)) {
+      // Add a new control in the FormArray
+      this.checkboxes.push(new FormControl(event.target.value));
+    } else {
+      // Remove the control form the FormArray
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts:32]: "from"

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      this.checkboxes.controls.forEach((ctrl: FormControl, i) => {
+        if (ctrl.value === event.target.value) {
+          this.checkboxes.removeAt(i);
+          return;
+        }
+      });
+    }
+  }
+
+  isSelected(code): AbstractControl {
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts:42]: The `code` parameter should have a type. `: string`, presumably...?

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    if (this.caseField.value && Array.isArray(this.caseField.value)) {
+      const values: string[] = this.caseField.value;
+      values.forEach(value => {
+        this.checkboxes.push(new FormControl(value));
+      });
+    }
+    this.registerControl(this.checkboxes, true);
+  }
+
+  onCheckChange(event) {
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.component.ts:27]: The `event` parameter should have a type...

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      <legend>
+        <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
+        <span *ngIf="caseField.hint_text" class="form-hint">{{caseField.hint_text}}</span>
+        <span *ngIf="checkboxes.errors && checkboxes.touched" class="error-message">{{checkboxes.errors | ccdFirstError}}</span>
+      </legend>
+  
+      <ng-container *ngFor="let checkbox of caseField.list_items">
+  
+        <div class="multiple-choice">
+          <input class="form-control" id="{{ buildElementId(checkbox.code) }}" [name]="id()" type="checkbox" [value]="checkbox.code" [checked]="isSelected(checkbox.code)" (change)="onCheckChange($event)">
```
Comment (paul-graham) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.html:14]: Can `[name]="id()"` be replaced with `name="{{ id() }}"` like you've done for the `id` attribute? :)

File: src/shared/components/palette/palette.service.ts [hunk 9]
```diff
@@ -65,7 +66,9 @@ export class PaletteService {
       case 'MoneyGBP':
         return write ? WriteMoneyGbpFieldComponent : ReadMoneyGbpFieldComponent;
       case 'DynamicList':
-        return write ? WriteDynamicListFieldComponent : ReadDynamicListFieldComponent;
+         return write ? WriteDynamicListFieldComponent : ReadDynamicListFieldComponent;
```
Comment (paul-graham) [inline src/shared/components/palette/palette.service.ts:69]: Indenting is off.

**Review by paul-graham — CHANGES_REQUESTED** [review]

**Review by paul-graham — APPROVED** [review]
Thank you for making the requested changes, @phillip-whitaker, and answering my annoying questions. Happy to approve this now. 👍

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.scss [hunk 10]
```diff
@@ -0,0 +1,16 @@
+.multi-select-list-field-table {
+
+    tr>td {
+      padding: 5px 0 5px 0;
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.scss:4]: Does this require linting? Also I believe 5px 0 would suffice.

Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.scss:4]: Just 5px 0?

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts:81]: For readability - 'should render values ...

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+      fixture.detectChanges();
+      
+      const cells = de.queryAll($VALUES);
+
+      expect(cells.length).toEqual(VALUES.length);
+    });
+
+    it('render some values as a table with one row and one cell per value', () => {
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts:89]: 'should render

File: src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+  describe('Persistable readonly multi-select-list field', () => {
+    const FORM_GROUP: FormGroup = new FormGroup({});
+    const CASE_FIELD = new CaseField();
+    CASE_FIELD.id = FIELD_ID;
+    CASE_FIELD.label = 'X';
+    CASE_FIELD.field_type = FIELD_TYPE;
+    CASE_FIELD.value = VALUES;
+    CASE_FIELD.list_items = FIELD_LIST_ITEMS;
+
+    let FieldReadComponent = MockComponent({
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/read-dynamic-multi-select-list-field.component.spec.ts:129]: const?

File: src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.html [hunk 14]
```diff
@@ -0,0 +1,23 @@
+<div class="form-group bottom-30" [ngClass]="{'error': !checkboxes.valid && checkboxes.touched}" [id]="id()">
+
+    <fieldset>
+  
+      <legend>
+        <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-multi-select-list/write-dynamic-multi-select-list-field.html:6]: Being finicky again but for readability can class="form-label" be moved after the ngIf as per the next two lines

File: src/shared/components/palette/palette.service.ts [hunk 15]
```diff
@@ -65,7 +66,9 @@ export class PaletteService {
       case 'MoneyGBP':
         return write ? WriteMoneyGbpFieldComponent : ReadMoneyGbpFieldComponent;
       case 'DynamicList':
-        return write ? WriteDynamicListFieldComponent : ReadDynamicListFieldComponent;
+        // return write ? WriteDynamicListFieldComponent : ReadDynamicListFieldComponent;
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/palette.service.ts:69]: Remove commented out code if not required

**Review by paulhowes-HMCTS — CHANGES_REQUESTED** [review]
A few finicky amends

**Review by paulhowes-HMCTS — DISMISSED** [review]
Fine now comments have been resolved

**Comment by Kartheek-S** [comment]
Changes are there in PR 1174 and this PR has no package file changes. So, closing this PR
